### PR TITLE
Ali/amp

### PR DESCRIPTION
**Pull request by AliGhadirzadeh**
AMP implementation

### CODE REVIEW

File: emote/algorithms/amp.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+class DiscriminatorLoss(LossCallback):
+    """
+    This loss is used to train a discriminator for
+    adversarial training.
+    """
+
+    def __init__(
+        self,
+        discriminator: nn.Module,
+        state_map_fn: Callable[[Tensor], Tensor],
```
Comment (klashenriksson) [inline emote/algorithms/amp.py:42]: I think we should split this up into imitation_state_map_fn and policy_state_map_fn, as they might require different mappings (this is the case for my work for instance)

File: emote/algorithms/amp.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        imitation_batch_size = imitation_batch["batch_size"]
+        policy_data_batch_size = policy_batch["batch_size"]
+
+        pos_obs: Tensor = imitation_batch["observation"][self._obs_key]
+        pos_next_obs: Tensor = imitation_batch["next_observation"][self._obs_key]
+        neg_obs: Tensor = policy_batch["observation"][self._obs_key]
+        neg_next_obs: Tensor = policy_batch["next_observation"][self._obs_key]
+
+        (pos_obs,) = (self._state_map_function(pos_obs),)
```
Comment (klashenriksson) [inline emote/algorithms/amp.py:83]: why the tuple wrap?

File: emote/memory/memory.py [hunk 3]
```diff
@@ -486,6 +486,30 @@ def __iter__(self):
             yield {self.data_group: data, self.size_key: data[self.size_key]}
 
 
+class MemoryLoaderList:
```
Comment (klashenriksson) [inline emote/memory/memory.py:489]: Please see my pull request here: https://github.com/EmbarkStudios/emote/pull/181 . The current implementation risks clobbering data if multiple loaders have the same datagroup, and there is also some troubles with the size_key that each loader adds.

**Review by klashenriksson — CHANGES_REQUESTED** [review]

File: emote/memory/memory.py (hunk 3 above)
Comment (AliGhadirzadeh) [reply emote/memory/memory.py:489]: Nice! I removed it from my PR then :)

File: emote/algorithms/amp.py (hunk 2 above)
Comment (AliGhadirzadeh) [reply emote/algorithms/amp.py:83]: Have no idea how this happened! It should be some automatic code formatting. Thanks for the catch :)

File: emote/algorithms/amp.py (hunk 1 above)
Comment (AliGhadirzadeh) [reply emote/algorithms/amp.py]: Fixed :)

File: emote/algorithms/amp.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        discriminator: nn.Module,
+        state_map_fn: Callable[[Tensor], Tensor],
+        style_reward_weight: float,
+        rollout_length: int,
+        observation_key: str,
+        data_group: str,
+    ):
+        super().__init__()
+        self._discriminator = discriminator
+        self._order = 0
```
Comment (klashenriksson) [inline emote/algorithms/amp.py:134]: Can this become a problem when paired with SAC? If QTarget runs before this callback it will miss the amp reward. Perhaps order just be an input parameter?

Comment (klashenriksson) [reply emote/algorithms/amp.py:134]: What happens if QTarget also has order=0?

Comment (klashenriksson) [reply emote/algorithms/amp.py:134]: Alright, I think this is fine then for this PR, users will have to assert that it runs before QTarget (or any other relevant callback).

I think I prefer to have the constructors take in an `order` parameter rather than having a `set_order` method. The callbacks are sorted by their order only once when given to the `Trainer`, so if you would call `set_order` at a later time, it would not have any effect on the actual training. By only exposing it during construction makes it clear that this is a "set once, never change" thing. A `get_order` method sounds nice though, we could make it a `@property`.

Anyway, that sounds like a different PR, approving this, nice stuff!

**Review by klashenriksson — APPROVED** [review]

**Review by singhblom — APPROVED** [review]
